##############
#README
#The script takes inter-parental genetic distance generated by PLINK using filtered SNPs,
#and correlate genetic distance of sub genelists with size
################
#Config & Data input
setwd("/ebio/abt6_projects7/SHB2/data/1_analysis/10_Subgenome_Genetdist/") #working directory
trios<-read.table("/ebio/abt6_projects7/SHB2/data/1_analysis/6_BTHdiffgenes/run197_BTHdiffgene_MetaRowIdx_forpairedMPHcalc_manualcorr.txt",header=T,sep="\t") #parental info
#read in plink 1001 distance matrix and id file. genetic distance calculated with 1kb promoter + gene body SNPs
distidls<-list()
  distidls[[1]]<-read.table("sub1001_AllBTHneg_1kbpromoter_hammingdist.mdist.id",header=F,sep="\t") #all BTH negative
  distidls[[2]]<-read.table("sub1001_pospos_1kbpromoter_hammingdist.mdist.id",header=F,sep="\t") #positive-positive
  distidls[[3]]<-read.table("sub1001_additiveOverlap_1kbpromoter_hammingdist.mdist.id",header=F,sep="\t")  #common additive genes
  distidls[[4]]<-read.table("wholegenome1001_hammingdist.mdist.id",header=F,sep="\t") #whole genome SNPs

  plinkdistls<-list()
  plinkdistls[[1]]<-read.table("sub1001_AllBTHneg_1kbpromoter_hammingdist.mdist",header=F,sep="\t")
  plinkdistls[[2]]<-read.table("sub1001_pospos_1kbpromoter_hammingdist.mdist",header=F,sep="\t")
  plinkdistls[[3]]<-read.table("sub1001_additiveOverlap_1kbpromoter_hammingdist.mdist",header=F,sep="\t")
  plinkdistls[[4]]<-read.table("wholegenome1001_hammingdist.mdist",header=F,sep="\t")
##################
  
  
#calculate average MPH of the trios
uniquetrio<-unique(trios[,c("FemaleGt","MaleGt")])
meantrioMPHfrac_Mock<-apply(uniquetrio,1,function(x){mean(trios[trios$FemaleGt%in%x[1] & trios$MaleGt%in%x[2] & trios$Treatment=="Mock",]$MPHfrac)}) 
meantrioMPHfrac_BTH<-apply(uniquetrio,1,function(x){mean(trios[trios$FemaleGt%in%x[1] & trios$MaleGt%in%x[2] & trios$Treatment=="BTH",]$MPHfrac)}) 
meantrioMPHmm2_Mock<-apply(uniquetrio,1,function(x){mean(trios[trios$FemaleGt%in%x[1] & trios$MaleGt%in%x[2] & trios$Treatment=="Mock",]$MPHmm2)}) 
meantrioMPHmm2_BTH<-apply(uniquetrio,1,function(x){mean(trios[trios$FemaleGt%in%x[1] & trios$MaleGt%in%x[2] & trios$Treatment=="BTH",]$MPHmm2)})

#match trio parents with plink matrix, subset
subdistls<-list()
for(i in 1:4){
  distid<-apply(distidls[[i]], c(1,2), function(x){paste0("p",x)})
  plinkdist<-plinkdistls[[i]]
  distidx<-t(apply(uniquetrio,1,function(x){return(c(which(distid[,1]==x[1]), which(distid[,2]==x[2])))}))
  subdistls[[i]]<-apply(distidx,1,function(x){plinkdist[x[1],x[2]]})
}

#plot genetic distance vs plant size
pdf("subgenome_hammingdist_1kbpromoter_MPH.pdf",height=14,width=14)
par(mfrow=c(2,2))

#AllBTH negative
plot(subdistls[[1]],meantrioMPHmm2_Mock, pch=19,col="springgreen4",xlim=c(0.15,0.5),ylim=c(-40,140),
     xlab="hamming distance (fraction)",ylab="Rosette MPH (mm2)",main="All BTH negative genes")
fit<-lm(meantrioMPHmm2_Mock~subdistls[[1]])
abline(a=fit$coefficients[1],b=fit$coefficients[2],col="springgreen4",lwd=3)
text(x=0.48,y=fit$coefficients[2]*0.48+fit$coefficients[1],paste0("R2=",round(summary(fit)$r.squared,digits=3)),pos=3,col="springgreen4")
points(subdistls[[1]],meantrioMPHmm2_BTH, pch=19,col="darkorange2")
fit<-lm(meantrioMPHmm2_BTH~subdistls[[1]])
abline(a=fit$coefficients[1],b=fit$coefficients[2],col="darkorange2",lwd=3)
text(x=0.18,y=fit$coefficients[2]*0.18+fit$coefficients[1],paste0("R2=",round(summary(fit)$r.squared,digits=3)),pos=1,col="darkorange2")

#positive-positive
plot(subdistls[[2]],meantrioMPHmm2_Mock, pch=19,col="springgreen4",xlim=c(0.15,0.5),ylim=c(-40,140),
     xlab="hamming distance (fraction)",ylab="Rosette MPH (mm2)",main="Positive-positive genes")
fit<-lm(meantrioMPHmm2_Mock~subdistls[[2]])
abline(a=fit$coefficients[1],b=fit$coefficients[2],col="springgreen4",lwd=3)
text(x=0.48,y=fit$coefficients[2]*0.48+fit$coefficients[1],paste0("R2=",round(summary(fit)$r.squared,digits=3)),pos=3,col="springgreen4")
points(subdistls[[2]],meantrioMPHmm2_BTH, pch=19,col="darkorange2")
fit<-lm(meantrioMPHmm2_BTH~subdistls[[2]])
abline(a=fit$coefficients[1],b=fit$coefficients[2],col="darkorange2",lwd=3)
text(x=0.18,y=fit$coefficients[2]*0.18+fit$coefficients[1],paste0("R2=",round(summary(fit)$r.squared,digits=3)),pos=1,col="darkorange2")

#Additive
plot(subdistls[[3]],meantrioMPHmm2_Mock, pch=19,col="springgreen4",xlim=c(0.15,0.5),ylim=c(-40,140),
     xlab="hamming distance (fraction)",ylab="Rosette MPH (mm2)",main="Positive-positive genes")
fit<-lm(meantrioMPHmm2_Mock~subdistls[[3]])
abline(a=fit$coefficients[1],b=fit$coefficients[2],col="springgreen4",lwd=3)
text(x=0.48,y=fit$coefficients[2]*0.48+fit$coefficients[1],paste0("R2=",round(summary(fit)$r.squared,digits=3)),pos=3,col="springgreen4")
points(subdistls[[3]],meantrioMPHmm2_BTH, pch=19,col="darkorange2")
fit<-lm(meantrioMPHmm2_BTH~subdistls[[3]])
abline(a=fit$coefficients[1],b=fit$coefficients[2],col="darkorange2",lwd=3)
text(x=0.18,y=fit$coefficients[2]*0.18+fit$coefficients[1],paste0("R2=",round(summary(fit)$r.squared,digits=3)),pos=1,col="darkorange2")

#Whole genome
plot(subdistls[[4]],meantrioMPHmm2_Mock, pch=19,col="springgreen4",xlim=c(0.15,0.5),ylim=c(-40,140),
     xlab="hamming distance (fraction)",ylab="Rosette MPH (mm2)",main="Whole Genome")
fit<-lm(meantrioMPHmm2_Mock~subdistls[[4]])
abline(a=fit$coefficients[1],b=fit$coefficients[2],col="springgreen4",lwd=3)
text(x=0.48,y=fit$coefficients[2]*0.48+fit$coefficients[1],paste0("R2=",round(summary(fit)$r.squared,digits=3)),pos=3,col="springgreen4")
points(subdistls[[4]],meantrioMPHmm2_BTH, pch=19,col="darkorange2")
fit<-lm(meantrioMPHmm2_BTH~subdistls[[4]])
abline(a=fit$coefficients[1],b=fit$coefficients[2],col="darkorange2",lwd=3)
text(x=0.18,y=fit$coefficients[2]*0.18+fit$coefficients[1],paste0("R2=",round(summary(fit)$r.squared,digits=3)),pos=1,col="darkorange2")

dev.off()
#########

#remove the one outlier that has small genetic distance
ol<-which(subdistls[[4]]<0.25) #since the genetic distance are highly correlated, it is the same outlier ID
#calculate pearson correlation value
AllBTHnegcor<-cor.test(subdistls[[1]],meantrioMPHmm2_Mock,method="pearson")
posposcor<-cor.test(subdistls[[2]],meantrioMPHmm2_Mock,method="pearson")
Additivecor<-cor.test(subdistls[[3]],meantrioMPHmm2_Mock,method="pearson")
Wholegncor<-cor.test(subdistls[[4]],meantrioMPHmm2_Mock,method="pearson")
AllBTHnegcor.bth<-cor.test(subdistls[[1]],meantrioMPHmm2_BTH,method="pearson")
posposcor.bth<-cor.test(subdistls[[2]],meantrioMPHmm2_BTH,method="pearson")
Additivecor.bth<-cor.test(subdistls[[3]],meantrioMPHmm2_BTH,method="pearson")
Wholegncor.bth<-cor.test(subdistls[[4]],meantrioMPHmm2_BTH,method="pearson")

pdf("subgenome_hammingdist_1kbpromoter_MPH_rmOL.pdf",height=14,width=14)
par(mfrow=c(2,2))
#AllBTH negative
plot(subdistls[[1]][-ol],meantrioMPHmm2_Mock[-ol], pch=19,col="springgreen4",xlim=c(0.25,0.5),ylim=c(-40,140),
     xlab="hamming distance (fraction)",ylab="Rosette MPH (mm2)",main="All BTH negative genes",axes = F)
axis(side=1,at=seq(0.25,0.5,0.05))
axis(side=2,at=seq(-40,140,20),las=2)
fit<-lm(meantrioMPHmm2_Mock[-ol]~subdistls[[1]][-ol])
abline(a=fit$coefficients[1],b=fit$coefficients[2],col="springgreen4",lwd=3)
text(x=0.48,y=fit$coefficients[2]*0.48+fit$coefficients[1]+20,
     paste0("R=",round(AllBTHnegcor$estimate,digits=3),", p=",round(AllBTHnegcor$p.value,digits=3)),
     pos=3,col="springgreen4")
points(subdistls[[1]][-ol],meantrioMPHmm2_BTH[-ol], pch=19,col="darkorange2")
fit<-lm(meantrioMPHmm2_BTH[-ol]~subdistls[[1]][-ol])
abline(a=fit$coefficients[1],b=fit$coefficients[2],col="darkorange2",lwd=3)
text(x=0.28,y=fit$coefficients[2]*0.28+fit$coefficients[1]+50,
     paste0("R=",round(AllBTHnegcor.bth$estimate,digits=3),", p=",round(AllBTHnegcor.bth$p.value,digits=3)),
     pos=1,col="darkorange2")

#positive-positive
plot(subdistls[[2]][-ol],meantrioMPHmm2_Mock[-ol], pch=19,col="springgreen4",xlim=c(0.25,0.5),ylim=c(-40,140),
     xlab="hamming distance (fraction)",ylab="Rosette MPH (mm2)",main="Positive-positive genes",axes=F)
axis(side=1,at=seq(0.25,0.5,0.05))
axis(side=2,at=seq(-40,140,20),las=2)
fit<-lm(meantrioMPHmm2_Mock[-ol]~subdistls[[2]][-ol])
abline(a=fit$coefficients[1],b=fit$coefficients[2],col="springgreen4",lwd=3)
text(x=0.48,y=fit$coefficients[2]*0.48+fit$coefficients[1]+20,
     paste0("R=",round(posposcor$estimate,digits=3),", p=",round(posposcor$p.value,digits=3)),
     pos=3,col="springgreen4")
points(subdistls[[2]][-ol],meantrioMPHmm2_BTH[-ol], pch=19,col="darkorange2")
fit<-lm(meantrioMPHmm2_BTH[-ol]~subdistls[[2]][-ol])
abline(a=fit$coefficients[1],b=fit$coefficients[2],col="darkorange2",lwd=3)
text(x=0.28,y=fit$coefficients[2]*0.28+fit$coefficients[1]+50,
     paste0("R=",round(posposcor.bth$estimate,digits=3),", p=",round(posposcor.bth$p.value,digits=3)),
     pos=1,col="darkorange2")

#Additive
plot(subdistls[[3]][-ol],meantrioMPHmm2_Mock[-ol], pch=19,col="springgreen4",xlim=c(0.25,0.5),ylim=c(-40,140),
     xlab="hamming distance (fraction)",ylab="Rosette MPH (mm2)",main="Additive genes",axes=F)
axis(side=1,at=seq(0.25,0.5,0.05))
axis(side=2,at=seq(-40,140,20),las=2)
fit<-lm(meantrioMPHmm2_Mock[-ol]~subdistls[[3]][-ol])
abline(a=fit$coefficients[1],b=fit$coefficients[2],col="springgreen4",lwd=3)
text(x=0.48,y=fit$coefficients[2]*0.48+fit$coefficients[1]+20,
     paste0("R=",round(Additivecor$estimate,digits=3),", p=",round(Additivecor$p.value,digits=3)),
     pos=3,col="springgreen4")
points(subdistls[[3]][-ol],meantrioMPHmm2_BTH[-ol], pch=19,col="darkorange2")
fit<-lm(meantrioMPHmm2_BTH[-ol]~subdistls[[3]][-ol])
abline(a=fit$coefficients[1],b=fit$coefficients[2],col="darkorange2",lwd=3)
text(x=0.28,y=fit$coefficients[2]*0.28+fit$coefficients[1]+50,
     paste0("R=",round(Additivecor.bth$estimate,digits=3),", p=",round(Additivecor.bth$p.value,digits=3)),
     pos=1,col="darkorange2")

#Whole genome
plot(subdistls[[4]][-ol],meantrioMPHmm2_Mock[-ol], pch=19,col="springgreen4",xlim=c(0.25,0.5),ylim=c(-40,140),
     xlab="hamming distance (fraction)",ylab="Rosette MPH (mm2)",main="Whole Genome",axes=F)
axis(side=1,at=seq(0.25,0.5,0.05))
axis(side=2,at=seq(-40,140,20),las=2)
fit<-lm(meantrioMPHmm2_Mock[-ol]~subdistls[[4]][-ol])
abline(a=fit$coefficients[1],b=fit$coefficients[2],col="springgreen4",lwd=3)
text(x=0.48,y=fit$coefficients[2]*0.48+fit$coefficients[1]+20,
     paste0("R=",round(Wholegncor$estimate,digits=3),", p=",round(Wholegncor$p.value,digits=3)),
     pos=3,col="springgreen4")
points(subdistls[[4]][-ol],meantrioMPHmm2_BTH[-ol], pch=19,col="darkorange2")
fit<-lm(meantrioMPHmm2_BTH[-ol]~subdistls[[4]][-ol])
abline(a=fit$coefficients[1],b=fit$coefficients[2],col="darkorange2",lwd=3)
text(x=0.28,y=fit$coefficients[2]*0.28+fit$coefficients[1]+50,
     paste0("R=",round(Wholegncor.bth$estimate,digits=3),", p=",round(Wholegncor.bth$p.value,digits=3)),
     pos=1,col="darkorange2")

dev.off()


########
#plot the association of whole-genome and sub-genome genetic distance
#histogram (density plot) of genetic distance across 1001 when using different subset
pdf("Hammingdist_distribution_by_genelist_1kbpromoter.pdf",height=14,width=14)
par(mfrow=c(2,1))
plot(density(unlist(plinkdistls[[4]])),col="grey50",lwd=4, 
     main="Hamming distance across 1001 by gene list", 
     xlab="Hamming distance (fraction)", 
     ylab="Density")
lines(density(unlist(plinkdistls[[3]])),col="indianred2",lwd=3)
lines(density(unlist(plinkdistls[[2]])),col="forestgreen",lwd=3)
lines(density(unlist(plinkdistls[[1]])),col="cornflowerblue",lwd=3)
legend(0.05,14,legend = c("Whole Genome","Additive","Positive-Positive","All BTH negative"),
       lwd=3,col=c("grey50","indianred2","cornflowerblue","forestgreen"))

plot(density(unlist(subdistls[[4]])),col="grey50",lwd=4, 
     main="Hamming distance across SHB2 by gene list", 
     xlab="Hamming distance (fraction)", 
     ylab="Density")
lines(density(unlist(subdistls[[3]])),col="indianred2",lwd=3)
lines(density(unlist(subdistls[[2]])),col="forestgreen",lwd=3)
lines(density(unlist(subdistls[[1]])),col="cornflowerblue",lwd=3)
legend(0.21,20,legend = c("Whole Genome","Additive","Positive-Positive","All BTH negative"),
       lwd=3,col=c("grey50","indianred2","cornflowerblue","forestgreen"))


dev.off()








